keps/sig-apps/20190318-PodDisruptionBudget-graduation-to-stable.md

```diff
@@ -0,0 +1,142 @@
+
+#### PDB in Deployment controller
+
+[Rolling update](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/rolling.go) of Deployment controller reads `MinAvailable` and `MaxUnavailable` from
+the rolling update spec and currently does not take PDB into account. We need to
+change the implementation and use
+`min(PDB.MinAvailable, RollingUpdate.MinAvailable)` and
+`max(PDB.MaxUnavailable, RollingUpdate.MaxUnavailable)` instead.
+
+### Risks and Mitigations
+
+We plan to support mutation of PDB objects that didn't exist in previous versions.
+The following needs to be checked and verified before graduation of the API.
+
+- [ ] Ensure that components do not have any logic that relies on immutability
+of PDBs. For example, if component builds a cache of pods that match various
+PDBs, they must add logic to invalid the cache on updates.
+   - Action Item: sweep for in-tree uses to make sure we’re informer driven and
+   aren’t assuming immutability.
+- [ ] Check what watches PDBs and make sure there is no performance concerns.
+- [ ] Updating a PDB could cause the state of a cluster to seem incorrect. For
+example, a PDB states that at least 10 replicas of a collection must be alive.
+Kubernetes control plane evicts some of the existing pods, but keeps at least 10
+around. The PDB is updated and states that at least 20 replicas must be kept
+alive. It may appear to an observer that the evictions happened before the PDB 
+update were incorrect, if they don’t notice the PDB update.
+  - Action Item: Update documents and explain mutablility of PDB its side effects.
+
+### Test Plan
+
+#### Existing Tests
+PodDisruptionBudget currently has tests in various components that use the feature:
+
+* Disruption controller
+  - [This test](https://github.com/kubernetes/kubernetes/blob/687d759e362b05dcdf11e336e2799704918e048d/pkg/controller/disruption/disruption_test.go#L140)
+  tests PDB MinAvailable, MaxUnavailable, and selector functionality in Disruption controller.
+* Kubectl
+  - [This test](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/generate/versioned/pdb_test.go)
+  tests generation of a PDB objects out of given parameters
+  - [This test](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/create/create_pdb_test.go)
+  tests creation of PDB objects from cmd parameters
+* Scheduler
+  - [This test](https://github.com/kubernetes/kubernetes/blob/ac56bd502ab96696682c66ebdff94b6e52471aa3/test/integration/scheduler/preemption_test.go#L731)
+  tests effects of PDB on preemption (PDB is honored in a best effort way)
+* Eviction integration tests
+  - [These tests](https://github.com/kubernetes/kubernetes/blob/master/test/integration/evictions/evictions_test.go) test eviction logic and its interactions with PDB.
+
+#### Needed Tests
+
+If this is considered a non-optional feature, there should be a conformance test
+for it (this needs to be an e2e test tagged as conformance).
```

Comment:
This seems like a good candidate for conformance testing. With an ack from @kubernetes/cncf-conformance-wg I'd make this a more definitive statement

Reply:
+1

Reply:
Changed wording to make the test a requirement.